proof/crefine/RISCV64/Finalise_C.thy

```diff
@@ -35,102 +35,226 @@ declare if_split [split del]
+  apply (rule allI, rule conseqPre)
+  apply (vcg exspec=rescheduleRequired_tcb_fields
+             exspec=thread_state_ptr_set_tsType_tcb_fields
+             exspec=isSchedulable_tcb_fields)
+  apply clarsimp
+  done
+
+lemma refill_head_overlapping_tcb_fields:
+  "\<forall>s. \<Gamma> \<turnstile>\<^bsub>/UNIV\<^esub> {s}
+       Call refill_head_overlapping_'proc
+       {s'. option_map2 tcbEPNext_C (cslift s') = option_map2 tcbEPNext_C (cslift s)
+            \<and> option_map2 tcbEPPrev_C (cslift s') = option_map2 tcbEPPrev_C (cslift s)
+            \<and> option_map2 tcbPriority_C (cslift s') = option_map2 tcbPriority_C (cslift s)
+            \<and> option_map2 tcbDomain_C (cslift s') = option_map2 tcbDomain_C (cslift s)}"
+  apply (rule allI, rule conseqPre)
+  apply vcg
+  apply clarsimp
+  done
+
+lemma refill_pop_head_tcb_fields:
+  "\<forall>s. \<Gamma> \<turnstile>\<^bsub>/UNIV\<^esub> {s}
+       Call refill_pop_head_'proc
+       {s'. option_map2 tcbEPNext_C (cslift s') = option_map2 tcbEPNext_C (cslift s)
+            \<and> option_map2 tcbEPPrev_C (cslift s') = option_map2 tcbEPPrev_C (cslift s)
+            \<and> option_map2 tcbPriority_C (cslift s') = option_map2 tcbPriority_C (cslift s)
+            \<and> option_map2 tcbDomain_C (cslift s') = option_map2 tcbDomain_C (cslift s)}"
+  apply (rule allI, rule conseqPre)
+  apply (vcg exspec=refill_head_modifies exspec=refill_next_modifies exspec=refill_size_modifies)
+  apply (clarsimp simp: typ_heap_simps)
+  done
+
+lemma merge_overlapping_head_refill_tcb_fields:
+  "\<forall>s. \<Gamma> \<turnstile>\<^bsub>/UNIV\<^esub> {s}
+       Call merge_overlapping_head_refill_'proc
+       {s'. option_map2 tcbEPNext_C (cslift s') = option_map2 tcbEPNext_C (cslift s)
+            \<and> option_map2 tcbEPPrev_C (cslift s') = option_map2 tcbEPPrev_C (cslift s)
+            \<and> option_map2 tcbPriority_C (cslift s') = option_map2 tcbPriority_C (cslift s)
+            \<and> option_map2 tcbDomain_C (cslift s') = option_map2 tcbDomain_C (cslift s)}"
+  apply (rule allI, rule conseqPre)
+  apply (vcg exspec=refill_pop_head_tcb_fields exspec=refill_head_modifies
+             exspec=refill_index_modifies)
+  apply (clarsimp simp: typ_heap_simps)
+  done
+
+lemma refill_unblock_check_tcb_fields:
+  "\<forall>s. \<Gamma> \<turnstile>\<^bsub>/UNIV\<^esub> {s}
+       Call refill_unblock_check_'proc
+       {s'. option_map2 tcbEPNext_C (cslift s') = option_map2 tcbEPNext_C (cslift s)
+            \<and> option_map2 tcbEPPrev_C (cslift s') = option_map2 tcbEPPrev_C (cslift s)
+            \<and> option_map2 tcbPriority_C (cslift s') = option_map2 tcbPriority_C (cslift s)
+            \<and> option_map2 tcbDomain_C (cslift s') = option_map2 tcbDomain_C (cslift s)}"
+  (is "\<forall>s. \<Gamma> \<turnstile>\<^bsub>/UNIV\<^esub> {s} ?f ({s'. ?post s s'})")
   apply (rule allI, rule conseqPre)
    apply vcg_step
    apply vcg_step
     apply vcg_step
    apply vcg_step
```

Comment:
something isn't right about this, but I also don't know what you're trying to do; would something like `apply (hoare_rule HoarePartial.ProcNoRec1)` help here? I definitely don't want this going in without some comments on why it's like this or what it's trying to do

Reply:
This is the most sensitive of the `hoarep/vcg` rules that I have in this PR, because `refill_unblock_check` contains a `While` loop. I use the `vcg_step` steps to very carefully get to the `While` loop, and then perform the `HoarePartial.reannotateWhileNoGuard` step. The other uses of this rule occur within `ccorres` rules where the stepping is done by `ccorres` split rules, I think. 

Similar use of `vcg_step` was made before in `setThreadState_cslift_spec`, which has been removed.

And yes, `apply (hoare_rule HoarePartial.ProcNoRec1)` will help cut down on these steps a little bit, but I can't completely remove the uses of `vcg_step`.

Would a comment after the `apply (hoare_rule HoarePartial.ProcNoRec1)` saying `(* step through the function until you get to the While loop *)` be reasonable?

And to answer the question of why all these `vcg` rules are needed, `cancelBadgedSends` has a `for` loop that ends up as a `While` loop of some kind in the Simpl. The proof is a rather disgusting induction (the step that performs the induction uses the rule `ccorres_inst_voodoo`) and the induction requires the body of the loop to not alter the head and the end of the queue over which the `for` loop runs. The previous rule `setThreadState_cslift_spec` established this for the much smaller operations which `master` does inside these loops. However, the functions performed in the loop for MCS are much more complicated (for example, `refill_unblock_check`, with its while loop) so we wanted to be very careful with this. With lots of help from Gerwin, we figured out that the preconditions of these `hoarep` rules could in fact be `UNIV`, which meant that the rather heavy-handed rule `tcbSchedEnqueue_cslift_precond_discharge` could be removed. We did need to think quite hard to do this, though, which is why, for instance, `thread_state_ptr_set_tcbQueued_tcb_fields` has some careful reasoning involving things such as `packed_heap_update_collapse_hrs`. There are so many rules because the functions called within the `for` loop contain many smaller functions, and we did not want to let `vcg` run wild and inline a whole bunch of things, just because that gives us terrible goals, and also takes a very long time in comparison with these verbose but fast proofs.

I was secretly hoping that these terrible vcg concerns would mean that we needed to handle the endpoint and notification queues with linked lists in the Haskell, similar to what I did for the ready and release queues. Then all this would be unnecessary, though of course an awful lot of work would be needed in Refine to carry that out. I think that these vcg proofs worked out really well in the end, and I'm hopefully that they will be sufficient for all loops we do within "ipc cancel" functions.

Reply:
It is perfectly possibly to annotate the while loop before we call `vcg`, let me look up the rule for that.

Reply:
I stand corrected, I can't find a rule in Simple that is analogous to `whileLoop_add_inv` in the monad world. We should definitely prove one and use that instead of `reannotateWhileNoGuard`.

The idea is to make a rule that is the same as `reannotateWhileNoGuard`, but is an equality instead of an implication. The equality can then be used with `subst` on the entire function body term before you invoke `vcg`, and when `vcg` gets to the while loop it will see a correctly annotated one. That only works nicely when there is only one while loop in the body, but I think that is the case here (if there are is more than one loop, you need to instantiate the rule a bit more to select the right one).

Reply:
`reannotateWhileNoGuard` is about the entire `hoarep` statement, not just about rewriting the "while" bit, as we have in `whileLoop_add_inv`, so I wasn't sure how to use that. I did find a bit of inspiration somewhere else in CRefine after searching for `whileAnno`, and have come up with this, which at least avoids the use of `vcg_step`:

```
  "∀s. Γ ⊢⇘/UNIV⇙ {s}
       Call refill_unblock_check_'proc
       {s'. option_map2 tcbEPNext_C (cslift s') = option_map2 tcbEPNext_C (cslift s)
            ∧ option_map2 tcbEPPrev_C (cslift s') = option_map2 tcbEPPrev_C (cslift s)
            ∧ option_map2 tcbPriority_C (cslift s') = option_map2 tcbPriority_C (cslift s)
            ∧ option_map2 tcbDomain_C (cslift s') = option_map2 tcbDomain_C (cslift s)}"
  (is "∀s. Γ ⊢⇘/UNIV⇙ {s} ?f ({s'. ?post s s'})")
  apply (hoare_rule HoarePartial.ProcNoRec1)
  apply (clarsimp simp: whileAnno_def)
  apply (rule_tac I1="{t. ?post t s}" in subst[OF whileAnno_def])
  apply (vcg exspec=merge_overlapping_head_refill_tcb_fields exspec=refill_head_overlapping_tcb_fields)
    subgoal by (clarsimp simp: option_map2_def typ_heap_simps split: if_splits)
   apply fastforce
  apply fastforce
  done
```

The `vcg` step does inline a few things like `refill_index` and `isRoundRobin`, so it's not very fast. And it seems like my invariant does need to mention the `s` that became a metavariable after the clarsimp did the `allI` for the HOL forall. So maybe `rule_tac` is something is the best I can do here.

Does that seem reasonable? Or should there still be something nicer that can be done?

Reply:
That looks reasonable to me.

Reply:
Yes, agreed, that looks fine. Nice trick with using the `subst` rule there. If you want to prevent unfolding in `vcg` you could provide spec rules for those functions -- it's possible that providing the modifies rules as spec rules is enough already.